library(ggplot2)
library(gplots)
library(reshape2)
library(caret)
library(plotROC)
library(pROC)
library(WriteXLS)


#setwd("") #This needs to be set to $MY_FOLDER/Step05_Make_figures

##################
###MAIN FIGURES###
##################

######################################################
#Figure 2b: Volcano plot
#This is generated by "$MY_FOLDER/Step01_Differential_splicing_analysis/fisherBased_psiTest.R"
#The figure is saved at "$MY_FOLDER/Output_figures/volcano_fdr.pdf"
######################################################


######################################################
#Figure 2c: RNASeq vs RT-PCR
a=read.table("RNASeq_vs_RTPCR.txt",header=T)
#PSI changes for RNASeq are calculated as the average changes from matrix "$MY_FOLDER/Input_data/SEMatrix_15477onWT.txt"
#PSI changes for PCR are calculated by wet-lab
cor(a$RNASeq,a$RTPCR)
ggplot(a,aes(x=RNASeq,y=RTPCR))+geom_point()+geom_smooth(method=lm)+geom_noBG()
ggsave(file="../Output_figures/RNASeq_vs_RTPCR.pdf", width = 8.27/5, height = 11.69/5, units = "in")
######################################################


######################################################
#Figure 3a: top motif contribution for prediction
a=read.table("../Output_data/3C_motifs_infl_param38/table_target.txt")
a=a[,c(1,3,4)]
b=reshape(a,idvar = "V1",timevar = "V3",direction = "wide")
colnames(b)=c("Filter","Incl","Excl","Unchanged")
rownames(b)=paste0("Motif",b[,1])
b=b[,-1]
d=b[which(abs(b$Incl)>0.1|abs(b$Excl)>0.1|abs(b$Unchanged)>0.1),]
d=as.matrix(d)
my_palette = colorRampPalette(c("#40A529", "white", "#6A1586"))(n = 19)
heatmap.2(d,scale ="row",col=my_palette,trace="none",tracecol = NA)
######################################################


######################################################
#Figure 3b: positional importance of top motifs, XI2 regions only
#Genearated by Python script: /data/talkowski/dg520/projects/Basset/src/basset_motif_convolution_dadi.py
#The raw figure is saved at "$MY_FOLDER/Output_data/3C_positional_weight_param38/positional_infl_entropy_heatmap.pdf"
######################################################


######################################################
#Figure 3c: Splicing entropy of candidates in three classes
a=read.table("3C_entropy.txt",header=T)
#"3C_entropy.txt" are calculated at 
#http://hollywood.mit.edu/burgelab/maxent/Xmaxentscan_scoreseq.html
#and
#http://hollywood.mit.edu/burgelab/maxent/Xmaxentscan_scoreseq_acc.html
#using coresponding subsequences from "$MY_FOLDER/Output_data/3C.fa" generated in Step01
a$Response=factor(a$Response,levels=c("Inclusion","Exclusion","Stable"))
ggplot(a,aes(x=Region,y=Entropy,fill=Response))+geom_boxplot(outlier.shape = NA)+scale_fill_manual(values=c("red","blue","black"))+geom_noBG()

t.test(a[a$Response=="Inclusion" & a$Region=="R1","Entropy"],a[a$Response=="Stable" & a$Region=="R1","Entropy"])$p.value
t.test(a[a$Response=="Exclusion" & a$Region=="R1","Entropy"],a[a$Response=="Stable" & a$Region=="R1","Entropy"])$p.value
t.test(a[a$Response=="Inclusion" & a$Region=="R1","Entropy"],a[a$Response=="Exclusion" & a$Region=="R1","Entropy"])$p.value

t.test(a[a$Response=="Inclusion" & a$Region=="R2","Entropy"],a[a$Response=="Stable" & a$Region=="R2","Entropy"])$p.value
t.test(a[a$Response=="Exclusion" & a$Region=="R2","Entropy"],a[a$Response=="Stable" & a$Region=="R2","Entropy"])$p.value
t.test(a[a$Response=="Inclusion" & a$Region=="R2","Entropy"],a[a$Response=="Exclusion" & a$Region=="R2","Entropy"])$p.value

t.test(a[a$Response=="Inclusion" & a$Region=="R3","Entropy"],a[a$Response=="Stable" & a$Region=="R3","Entropy"])$p.value
t.test(a[a$Response=="Exclusion" & a$Region=="R3","Entropy"],a[a$Response=="Stable" & a$Region=="R3","Entropy"])$p.value
t.test(a[a$Response=="Inclusion" & a$Region=="R3","Entropy"],a[a$Response=="Exclusion" & a$Region=="R3","Entropy"])$p.value

t.test(a[a$Response=="Inclusion" & a$Region=="R4","Entropy"],a[a$Response=="Stable" & a$Region=="R4","Entropy"])$p.value
t.test(a[a$Response=="Exclusion" & a$Region=="R4","Entropy"],a[a$Response=="Stable" & a$Region=="R4","Entropy"])$p.value
t.test(a[a$Response=="Inclusion" & a$Region=="R4","Entropy"],a[a$Response=="Exclusion" & a$Region=="R4","Entropy"])$p.value
######################################################


######################################################
#Figure 3d-e: minigene validation to examine the robustness of the CNN model

#script to figure out thresholds of prediction (e.g. inclusion, exclusion and unchanged): $MY_FOLDER/Step02_Build_CNN_model/probThreshold.R
#inclusion cutoff: 0.225832171738145
#exclusion cutoff: 0.803149670362475
#unchanged cutoff: 0.36993284523487

normP=function(i,e,u){
  n1=i/0.225832171738145
  n2=e/0.803149670362475
  n3=u/0.36993284523487
  f1=n1/n3
  f2=n2/n3
  np1=f1/(f1+f2+1)
  np2=f2/(f1+f2+1)
  np3= 1/(f1+f2+1)
  d=data.frame(c("Inclusion","Exclusion","Unchanged"),c(np1,np2,np3))
  colnames(d)=c("Response","Normalized_probability")
  d$Response=factor(d$Response,levels=c("Inclusion","Exclusion","Unchanged"))
  return(d)
}

#For each of the following plot
#WT sequence is extracted by exon triplet name from "$MY_FOLDER/Output_data/3C.fa" which is generated in Step01
#MUT sequence is manually edited from WT sequence
#The prediction are made using the following command
#cd $BASSET_FOLDER/src/
#python basset_predict.py\
#    $MY_FOLDER/Output_data/3C_cnn_param38_best.th\
#    SEQUENCE.fa\
#    $MY_FOLDER/Output_data/SEQUENCE.fa.param38.prob
#The output file "$MY_FOLDER/Output_data/SEQUENCE.fa.param38.prob" contains three values
#representing prediction scores for inclusion, exclusion and unchanged category, respectively

#Figure 3d triplet name: ENSG00000168137@SETD5@3:+:9476170^9476273&9476315^9476507|9476170^9476507
#WT prediction: 
wt=normP(0.068559519946575, 0.64225089550018, 0.23061296343803)
ggplot(wt,aes(x=Response,y=Normalized_probability,fill=Response))+geom_bar(stat="identity")+geom_noBG()+geom_fill(c("red","blue","black"))+scale_y_continuous(expand=c(0,0))
#MUT:+6A>C prediction: 
mt=normP(0.052986100316048,0.72428953647614,0.18079993128777)
ggplot(mt,aes(x=Response,y=Normalized_probability,fill=Response))+geom_bar(stat="identity")+geom_noBG()+geom_fill(c("red","blue","black"))+scale_y_continuous(expand=c(0,0))

#Figure 3e triplet name: ENSG00000137817@PARP6@15:-:72542434^72543185&72543296^72543547|72542434^72543547
#WT prediction: 
wt=normP(0.035159323364496, 0.063608586788177,0.049906201660633)
ggplot(wt,aes(x=Response,y=Normalized_probability,fill=Response))+geom_bar(stat="identity")+geom_noBG()+geom_fill(c("red","blue","black"))+scale_y_continuous(expand=c(0,0))
#MUT:+6G>T prediction: 
mt=normP(0.34046351909637,0.2087162733078,0.37880665063858)
ggplot(mt,aes(x=Response,y=Normalized_probability,fill=Response))+geom_bar(stat="identity")+geom_noBG()+geom_fill(c("red","blue","black"))+scale_y_continuous(expand=c(0,0))
######################################################


######################################################
#Figure 4a: Step-wised filter down, from total clinvar pathogenic mutations to potential rescue
a=data.frame(c(264*100/118771,(14885-264)*100/118771,(19164-14885)*100/118771,(118771-19164)*100/118771),c("D","C","B","A"),"ClinVar")
colnames(a)=c("Percentage","Class","DB")
ggplot(a, aes(DB)) + geom_col(aes(y=Percentage,fill=Class))+geom_noBG()+geom_fill(c("grey","orange","darkgreen","red"))
ks.test(a[a$V2=="roi",1],a[a$V2=="bkg",1])
######################################################


######################################################
#Figure 4b: Violin plot of distantce to the closest annotated junction
a=read.table("dist.txt")
ggplot(a, aes(x=V2, y=log10(V1))) + geom_violin()+geom_noBG()+geom_hline(yintercept = log10(75),linetype="dashed")
ks.test(a[a$V2=="roi",1],a[a$V2=="bkg",1])
######################################################

######################################################
#Table 1. Potential therapeutic targets of BPN-15477 with top 20 allele frequencies
#Table 1 is a subset of Supplementary Table 3, which is generated at "$MY_FOLDER/Output_data/sad_table_rescue_alleleFreq_4supplementaryTable_wAltExon.txt"
######################################################


######################################################
#Figure 4c-f: Prediction on clinvar mutations
#For each of the following plot
#MUT sequence is extracted by exon triplet name from the last column of "$MY_FOLDER/Output_data/clinvar_20190325_SNV_INDEL1000_mainChr_headerLength_spliceAIpredicted_ds02_pathogenic_junctions_mEX_4BassetSad.vcf"
#generated by the script "$MY_FOLDER/Step03_Find_ClinVar_splicing_mutations/filterSpliceAIpredictedVCF4BassetSAD_param38.py"
#The prediction are made using the following command
#cd $BASSET_FOLDER/src/
#python basset_predict.py\
#    $MY_FOLDER/Output_data/3C_cnn_param38_best.th\
#    SEQUENCE.fa\
#    $MY_FOLDER/Output_data/SEQUENCE.fa.param38.prob
#The output file "$MY_FOLDER/Output_data/SEQUENCE.fa.param38.prob" contains three values
#representing prediction scores for inclusion, exclusion and unchanged category, respectively

#triplet name: ENSG00000107798@LIPA@10:-:90975767^90982267&90982340^90983440|90975767^90983440
#ClinVar ID: 203361
mt=normP(0.6312660574913,0.050920579582453,0.29948154091835)
ggplot(mt,aes(x=Response,y=Normalized_probability,fill=Response))+geom_bar(stat="identity")+geom_noBG()+geom_fill(c("red","blue","black"))+scale_y_continuous(expand=c(0,0))

#triplet name: ENSG00000001626@CFTR@7:+:117243837^117246727&117246808^117250572|117243837^117250572
#ClinVar ID: 35857
mt=normP(0.43889421224594, 0.15516199171543, 0.17330984771252)
ggplot(mt,aes(x=Response,y=Normalized_probability,fill=Response))+geom_bar(stat="identity")+geom_noBG()+geom_fill(c("red","blue","black"))+scale_y_continuous(expand=c(0,0))

#triplet name: ENSG00000076242@MLH1@3:+:37089175^37090007&37090101^37090394|37089175^37090394
#ClinVar ID: 89979
mt=normP(0.32855752110481, 0.28068980574608, 0.25360396504402)
ggplot(mt,aes(x=Response,y=Normalized_probability,fill=Response))+geom_bar(stat="identity")+geom_noBG()+geom_fill(c("red","blue","black"))+scale_y_continuous(expand=c(0,0))

#triplet name: MAPTT@17:+:44074031^44087675&44087769^44091608|44074031^44091608
#ClinVar ID: 98222
mt=normP(0.015885507687926,0.94923764467239,0.041594445705414)
ggplot(mt,aes(x=Response,y=Normalized_probability,fill=Response))+geom_bar(stat="identity")+geom_noBG()+geom_fill(c("red","blue","black"))+scale_y_continuous(expand=c(0,0))
######################################################


###########################
###SUPPLEMENTARY FIGURES###
###########################

######################################################
#Supplementary Table 2
#Generated by the script at "$MY_FOLDER/Step01_Differential_splicing_analysis/fisherBased_psiTest.R"
######################################################


######################################################
#Supplementary Figure 1b: training loss vs loss
loss0=read.table("trainProcess_3C_param38.txt",header=T)
#"trainProcess_3C_param38.txt" can be copied from standard output during the training of the CNN model
#by the following command
#cd $BASSET_FOLDER/src/
#./basset_train.lua\
#  -job $MY_FOLDER/Step02_Build_CNN_model/param38.txt\
#  -save $MY_FOLDER/Output_data/3C_cnn_param38\
#  -result $MY_FOLDER/Output_data/3C_cnn_loss_param38\
#  -rand 122\
#  $MY_FOLDER/Output_data/3C_learn.h5
auc=loss0[,c("Epoch","AUC")]
loss1=loss0[,c("Epoch","Train_loss")]
loss2=loss0[,c("Epoch","Valid_loss")]
colnames(loss1)=colnames(loss2)=c("Epoch","Loss")
loss=rbind(loss1,loss2)
loss$Group=c(rep("Training_loss",22),rep("Validation_loss",22))
ggplot(loss,aes(x=Epoch,y=Loss,color=Group))+geom_line()+geom_noBG()+scale_color_manual(values=c("red","blue"))+geom_vline(xintercept=12,linetype="dashed",color="grey")+geom_line(data=auc,aes(x=Epoch,y=AUC*2),color="black")+scale_y_continuous(name = "Loss", sec.axis = sec_axis(~./2, name = "Average AUC"))
ggsave("../Output_figures/trainProcess_3C_param38.pdf")
######################################################


######################################################
#Supplementary Figure 1c: ROC curves of the predictions on the test sets
#This is generated by the R script at "$MY_FOLDER/Step02_Build_CNN_model/probThreshold.R"
#The figure is saved at "$MY_FOLDER/Output_figures/aucs_param38.pdf"
######################################################


######################################################
#Supplementary Figure 2: all motif contribution for prediction
a=read.table("../Output_data/3C_motifs_infl_param38/table_target.txt")
a=a[,c(1,3,4)]
b=reshape(a,idvar = "V1",timevar = "V3",direction = "wide")
colnames(b)=c("Filter","Incl","Excl","Unchanged")
rownames(b)=paste0("Motif",b[,1])
b=b[,-1]
d=b[which(b$Incl=0 & abs(b$Excl)!=0 & abs(b$Unchanged)!=0),]
d=as.matrix(d)
my_palette <- colorRampPalette(c("#40A529", "white", "#6A1586"))(n = 19)
heatmap.2(d,scale ="row",col=my_palette,trace="none",tracecol = NA)
######################################################


######################################################
#Supplementary Figure 3a: positional importance of top motifs, all four regions
#Genearated by Python script "$BASSET_FOLDER/src/basset_motif_convolution_dadi.py" during Step02
#by the following command
#python basset_motif_convolution_dadi.py -s 131 -i\
# -o $MY_FOLDER/Output_data/3C_positional_weight_param38\
# $MY_FOLDER/Output_data/3C_cnn_param38_best.th\
# $MY_FOLDER/Output_data/3C_learn.h5
#The raw figure is saved at "$MY_FOLDER/Output_data/3C_positional_weight_param38/positional_infl_entropy_heatmap.pdf"
######################################################

######################################################
#Supplementary Figure 3c: minigene validation to examine the robustness of the CNN model
#For each of the following plot
#WT sequence is extracted by exon triplet name from "$MY_FOLDER/Output_data/3C.fa" which is generated in Step01
#MUT sequence is manually edited from WT sequence
#The prediction are made using the following command
#cd $BASSET_FOLDER/src/
#python basset_predict.py\
#    $MY_FOLDER/Output_data/3C_cnn_param38_best.th\
#    SEQUENCE.fa\
#    $MY_FOLDER/Output_data/SEQUENCE.fa.param38.prob
#The output file "$MY_FOLDER/Output_data/SEQUENCE.fa.param38.prob" contains three values
#representing prediction scores for inclusion, exclusion and unchanged category, respectively
#triplet name: ENSG00000149532@CPSF7@11:-:61188046^61188663&61188730^61188861|61188046^61188861
#WT prediction:
wt=normP(0.53450781106949, 0.15479181706905, 0.26954635977745)
ggplot(wt,aes(x=Response,y=Normalized_probability,fill=Response))+geom_bar(stat="identity")+geom_noBG()+geom_fill(c("red","blue","black"))+scale_y_continuous(expand=c(0,0),breaks=c(0,0.1,0.2,0.3,0.4,0.5,0.6))
#MUT:+6A>C prediction:
mt=normP(0.8227231502533,0.031591784209013,0.18625923991203)
ggplot(mt,aes(x=Response,y=Normalized_probability,fill=Response))+geom_bar(stat="identity")+geom_noBG()+geom_fill(c("red","blue","black"))+scale_y_continuous(expand=c(0,0))
######################################################


######################################################
#Supplementary Figure 4: Prediction on clinvar mutations
#For each of the following plot
#MUT sequence is extracted by exon triplet name from the last column of "$MY_FOLDER/Output_data/clinvar_20190325_SNV_INDEL1000_mainChr_headerLength_spliceAIpredicted_ds02_pathogenic_junctions_mEX_4BassetSad.vcf"
#generated by the script "$MY_FOLDER/Step03_Find_ClinVar_splicing_mutations/filterSpliceAIpredictedVCF4BassetSAD_param38.py"
#The prediction are made using the following command
#cd $BASSET_FOLDER/src/
#python basset_predict.py\
#    $MY_FOLDER/Output_data/3C_cnn_param38_best.th\
#    SEQUENCE.fa\
#    $MY_FOLDER/Output_data/SEQUENCE.fa.param38.prob
#The output file "$MY_FOLDER/Output_data/SEQUENCE.fa.param38.prob" contains three values
#representing prediction scores for inclusion, exclusion and unchanged category, respectively

#triplet name: MAPTT@17:+:44074031^44087675&44087769^44091608|44074031^44091608
#ClinVar ID: 98218
mt=normP(0.019421331584454, 0.93527942895889, 0.052945721894503)
ggplot(mt,aes(x=Response,y=Normalized_probability,fill=Response))+geom_bar(stat="identity")+geom_noBG()+geom_fill(c("red","blue","black"))+scale_y_continuous(expand=c(0,0))
######################################################